features/include/pcl/features/impl/good.hpp

```diff
@@ -0,0 +1,454 @@
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef PCL_FEATURES_IMPL_GOOD_H_
+#define PCL_FEATURES_IMPL_GOOD_H_
+#include <pcl/features/good.h>
+#include <pcl/filters/project_inliers.h>
+#include <pcl/point_types.h>
+
+
+#include <numeric> ///std::accumulate
+
+//////////////////////////////////////////////////////////////////////////////////////////////  
+
+template <typename PointInT, int BinN> 
+void 
+pcl::GOODEstimation<PointInT, BinN>::computeBoundingBoxDimensions ()
+{
+  PointInT minimum_pt, maximum_pt, dimensions;  
+  pcl::getMinMax3D (*transformed_point_cloud_, minimum_pt, maximum_pt); // min max for bounding box
+  dimensions.getVector3fMap () = maximum_pt.getVector3fMap () - minimum_pt.getVector3fMap ();
+  bbox_dimensions_ << dimensions.x, dimensions.y, dimensions.z;
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int BinN>
+void
+pcl::GOODEstimation<PointInT, BinN>::projectPointCloudToPlane (PointCloudInPtr pc_in, pcl::ModelCoefficients::Ptr coefficients, PointCloudInPtr pc_out)
+{
+  ///Create the projection object
+  pcl::ProjectInliers<PointInT> projection;
+  projection.setModelType (pcl::SACMODEL_NORMAL_PLANE); 
+  projection.setInputCloud (pc_in);
+  projection.setModelCoefficients (coefficients);
+  projection.filter (*pc_out);
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int BinN>
+void 
+pcl::GOODEstimation<PointInT, BinN>::convert2DHistogramTo1DHistogram (const std::vector<std::vector<unsigned int> >  histogram_2D, std::vector<unsigned int>  &histogram)
+{
+  for (size_t i = 0 ; i < BinN; i++)
+  {   
+    for (size_t j = 0; j < BinN; j++)
+    {
+      histogram.at (i * BinN + j) = histogram_2D.at(i).at(j);
+    }
+  }
+}
```

Comment:
- `histogram_2D` needs to be `const std::vector<std::vector<unsigned int> >&`
- Replace implementation for
```cpp
template<typename PointInT, int BinN> std::vector<unsigned>
pcl::GOODEstimation<PointInT, BinN>::convert2DHistogramTo1DHistogram (const std::vector<std::vector<unsigned> >& histogram_2D)
{
  std::vector<unsigned> histogram (BinN * BinN);
  for (size_t i = 0, j = 0; i < BinN; ++i, j += BinN)
    std::copy (histogram_2D[i].begin (), histogram_2D[i].end (), &histogram[j]);

  return histogram
}
```